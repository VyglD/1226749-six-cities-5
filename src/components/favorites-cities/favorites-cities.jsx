import React from "react";
import {connect} from "react-redux";
import OffersListFavorite from "../offers-list-favorite/offers-list-favorite";
import {
  offersType,
  offerIdsType,
} from "../../types";


const getFavoriteOffersByCities = (offers, ids) => {
  const favoriteOffersByCities = new Map();

  offers.filter((offer) => ids.includes(offer.id))
    .forEach((offer) => {
      if (favoriteOffersByCities.has(offer.city)) {
        favoriteOffersByCities.get(offer.city).push(offer);
      } else {
        favoriteOffersByCities.set(offer.city, [offer]);
      }
    });

  return favoriteOffersByCities;
};

const FavoritesCities = (props) => {
  const {
    allOffers,
    favoriteOfferIds,
  } = props;

  const offersByCities = getFavoriteOffersByCities(allOffers, favoriteOfferIds);

  return (
    <ul className="favorites__list">
      {
        Array.from(offersByCities.entries()).map(([city, offers]) => (
          <li className="favorites__locations-items" key={city}>
            <div className="favorites__locations locations locations--current">
              <div className="locations__item">
                <a className="locations__item-link" href="#">
                  <span>{city}</span>
                </a>
              </div>
            </div>
            <OffersListFavorite
              offers={offers}
              {...props}
            />
          </li>
        ))
      }
    </ul>
  );
};

FavoritesCities.propTypes = {
  allOffers: offersType,
  favoriteOfferIds: offerIdsType,
};

const mapStateToProps = ({OFFERS}) => ({
  allOffers: OFFERS.allOffers,
});

export {FavoritesCities};
export default connect(mapStateToProps)(FavoritesCities);
