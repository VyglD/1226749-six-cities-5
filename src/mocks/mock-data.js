const Key = {
  ENTER: `Enter`,
  ESC: `Escape`,
  SPACE: ` `,
  TAB: `Tab`,
  SHIFT: `Shift`,
  UP: `ArrowUp`,
  DOWN: `ArrowDown`,
};

const CITIES = [
  `Paris`,
  `Cologne`,
  `Brussels`,
  `Amsterdam`,
  `Hamburg`,
  `Dusseldorf`
];

const city = CITIES[3];
const activeCity = CITIES[0];

const HousingType = {
  APARTMENT: `Apartment`,
  ROOM: `Private Room`,
  HOUSE: `House`,
  HOTEL: `Hotel`
};

const SortType = {
  DEFAULT: {
    value: `Popular`,
    method: () => {},
  },
  PRICE_TO_HIGH: {
    value: `Price: low to high`,
    method: (i, j) => (i.cost - j.cost),
  },
  PRICE_TO_LOW: {
    value: `Price: high to low`,
    method: (i, j) => (j.cost - i.cost),
  },
  RATE: {
    value: `Top rated first`,
    method: (i, j) => (j.rate - i.rate),
  },
};

const mockOffers = [
  {
    id: `1`,
    city: CITIES[0],
    latitude: 48.95,
    longitude: 2.35,
    title: `Title of first server offer`,
    photos: [`img/photo1`, `img/photo11`],
    description: `Description of first server offer`,
    isPremium: true,
    housingType: HousingType.APARTMENT,
    rate: 4.5,
    rooms: 4,
    guests: 4,
    cost: 111,
    features: [`Washer`, `Towels`, `Fridge`],
    owner: {
      avatar: `img/avatar-angelina.jpg`,
      name: `Anna`,
      isSuper: true,
    },
    cityInfo: {
      name: CITIES[0],
      latitude: 48.85661,
      longitude: 2.351499,
      zoom: 12,
    }
  },
  {
    id: `2`,
    city: CITIES[0],
    latitude: 48.87,
    longitude: 2.33,
    title: `Title of cheapest offer`,
    photos: [],
    description: `Description of cheapest offer`,
    isPremium: false,
    housingType: HousingType.HOTEL,
    rate: 3,
    rooms: 3,
    guests: 3,
    cost: 1,
    features: [],
    owner: {
      avatar: `img/avatar-angelina.jpg`,
      name: `Angelina`,
      isSuper: true,
    },
    cityInfo: {
      name: CITIES[0],
      latitude: 48.85661,
      longitude: 2.351499,
      zoom: 12,
    }
  },
  {
    id: `3`,
    city: CITIES[1],
    latitude: 50.95,
    longitude: 6.95,
    title: `Title of most expensive offer`,
    photos: [`img/photo3`],
    description: `Description of most expensive offer`,
    isPremium: true,
    housingType: HousingType.APARTMENT,
    rate: 4.5,
    rooms: 4,
    guests: 4,
    cost: 4444,
    features: [`Washer`, `Towels`, `Fridge`],
    owner: {
      avatar: `img/avatar-angelina.jpg`,
      name: `Anna`,
      isSuper: true,
    },
    cityInfo: {
      name: CITIES[1],
      latitude: 50.938361,
      longitude: 6.959974,
      zoom: 13,
    },
  },
  {
    id: `4`,
    city: CITIES[2],
    latitude: 50.83,
    longitude: 4.34,
    title: `Title of top rated offer`,
    photos: [`img/photo4`, `img/photo44`],
    description: `Description of top rated offer`,
    isPremium: true,
    housingType: HousingType.HOUSE,
    rate: 5,
    rooms: 1,
    guests: 1,
    cost: 222,
    features: [`Washer`, `Towels`, `Fridge`],
    owner: {
      avatar: `img/avatar-angelina.jpg`,
      name: `Anna`,
      isSuper: true,
    },
    cityInfo: {
      name: CITIES[1],
      latitude: 50.846557,
      longitude: 4.351697,
      zoom: 13,
    },
  },
];

const mockOffersFavoriteIds = [`1`, `2`];

const mockReview = {
  id: `1`,
  name: `Bob`,
  photo: `img/photo`,
  rate: 3,
  date: new Date(0),
  text: `comment`,
};

const mockFunction = () => {};

const mockEvent = {
  preventDefault: mockFunction,
};

const mockEscKeyEvent = Object.assign(
    {},
    mockEvent,
    {
      key: Key.ESC,
    }
);

export {
  Key,
  city,
  activeCity,
  CITIES,
  SortType,
  mockOffers,
  mockOffersFavoriteIds,
  mockReview,
  mockFunction,
  mockEvent,
  mockEscKeyEvent,
};
