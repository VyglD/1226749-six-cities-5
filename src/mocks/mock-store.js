import configureStore from 'redux-mock-store';
import {
  mockOffers,
  mockOffersFavoriteIds,
  activeCity,
} from "./mock-data";

const createStore = configureStore();

const mockStore = createStore({
  OFFERS: {
    allOffers: mockOffers,
  },
  FAVORITES: {
    favoriteIds: mockOffersFavoriteIds,
  },
  OFFER: {
    openedOfferId: null,
    reviews: [],
    nearOffers: [],
  },
  USER: {
    isLogin: false,
    email: ``,
    avatar: ``,
    isSuper: false,
  },
  REDIRECT: {},
  CITY: {
    activeCity,
  },
});

export {mockStore};
