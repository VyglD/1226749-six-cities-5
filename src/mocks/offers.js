import {HOUSING_TYPE} from "../const";

const AVATAR_URL = `https://api.adorable.io/avatars/74`;
const BIG_NUMBER = 999;

const MAX_PHOTOS = 5;
const MAX_RATE = 5;
const MAX_ROOMS = 3;
const MAX_GUESTS = 4;

const features = [
  `Wi-Fi`,
  `Heating`,
  `Kitchen`,
  `Fridge`,
  `Washing machine`,
  `Coffee machine`,
  `Dishwasher`,
  `Towels`,
  `Baby seat`,
  `Cabel TV`
];

const getRandomInteger = (a = 0, b = 1) => {
  const lower = Math.ceil(Math.min(a, b));
  const upper = Math.floor(Math.max(a, b));

  return Math.floor(lower + Math.random() * (upper - lower + 1));
};

const generatePhotos = () => {
  return new Array(getRandomInteger(MAX_PHOTOS))
    .fill()
    .map(() => {
      return `https://picsum.photos/id/${getRandomInteger(BIG_NUMBER)}/260/200`;
    });
};

const getRandomElement = (array) => {
  return array[getRandomInteger(0, array.length - 1)];
};

const shuffleArray = (arr) => {
  return arr.slice().sort(() => {
    return 0.5 - Math.random();
  });
};

const getRandomSubArray = (array) => {
  const length = array.length;
  const half = Math.floor(length / 2);
  const start = getRandomInteger(0, half);
  const end = getRandomInteger(half, length);

  return shuffleArray(array).slice(start, end);
};

export default [
  {
    id: `1`,
    city: `Amsterdam`,
    title: `Beautiful & luxurious apartment at great location`,
    photos: [`/img/apartment-01.jpg`, ...generatePhotos()],
    description: [
      `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`
    ],
    isPremium: true,
    housingType: HOUSING_TYPE.APARTMENT,
    rate: getRandomInteger(MAX_RATE),
    rooms: getRandomInteger(MAX_ROOMS),
    guests: getRandomInteger(MAX_GUESTS),
    cost: 120,
    features: getRandomSubArray(features),
    owner: {
      avatar: `/img/avatar-angelina.jpg`,
      name: `Angelina`,
      isSuper: true
    }
  },
  {
    id: `2`,
    city: `Amsterdam`,
    title: `Wood and stone place`,
    photos: [`/img/room.jpg`, ...generatePhotos()],
    description: [
      `This average-sized, rectangular bedroom has matching wooden furniture.
        The floor is stone and the walls are painted and decorated with a wallpaper border.
        Light is provided by table lamps and a ceiling light.
        The room is done in colors that remind you of Easter and overall has a quirky look.
        Among the first things one notices walking in is a collection of action figures.`
    ],
    isPremium: false,
    housingType: HOUSING_TYPE.ROOM,
    rate: getRandomInteger(MAX_RATE),
    rooms: getRandomInteger(MAX_ROOMS),
    guests: getRandomInteger(MAX_GUESTS),
    cost: 80,
    features: getRandomSubArray(features),
    owner: {
      avatar: `${AVATAR_URL}/${Math.random()}`,
      name: `Mark`,
      isSuper: false
    }
  },
  {
    id: `3`,
    city: `Amsterdam`,
    title: `Canal View Prinsengracht`,
    photos: [`/img/apartment-02.jpg`, ...generatePhotos()],
    description: [
      `This spacious, square office has matching metal and plastic furniture.
        The floor is carpeted and the walls are painted with a paneled dado.
        Light is provided by wall lamps and a ceiling light.`,
      `The room is done in a vintage theme in warm colors and overall looks like an old castle.
        Among the first things one notices walking in is art supplies on the desk.`
    ],
    isPremium: false,
    housingType: getRandomElement(Object.values(HOUSING_TYPE)),
    rate: getRandomInteger(MAX_RATE),
    rooms: getRandomInteger(MAX_ROOMS),
    guests: getRandomInteger(MAX_GUESTS),
    cost: 132,
    features: getRandomSubArray(features),
    owner: {
      avatar: `${AVATAR_URL}/${Math.random()}`,
      name: `Julia`,
      isSuper: false
    }
  },
  {
    id: `4`,
    city: `Amsterdam`,
    title: `Nice, cozy, warm big bed apartment`,
    photos: [`/img/apartment-03.jpg`, ...generatePhotos()],
    description: [
      `This average-sized, square dining room has matching wooden furniture.`,
      `The seating is cushioned.`,
      `The floor is tiled and the walls are painted and decorated with a wallpaper border.`,
      `Light is provided by floor lamps and a ceiling light.`,
      `The room is done in colors that remind you of a coffee shop and overall looks very modern.
        Among the first things one notices walking in is a strange clock.`
    ],
    isPremium: true,
    housingType: getRandomElement(Object.values(HOUSING_TYPE)),
    rate: getRandomInteger(MAX_RATE),
    rooms: getRandomInteger(MAX_ROOMS),
    guests: getRandomInteger(MAX_GUESTS),
    cost: 180,
    features: getRandomSubArray(features),
    owner: {
      avatar: `${AVATAR_URL}/${Math.random()}`,
      name: `Tom`,
      isSuper: true
    }
  }
];
