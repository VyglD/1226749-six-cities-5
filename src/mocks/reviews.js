export default [
  {
    offerId: `1`,
    name: `Max`,
    rate: 4,
    date: `April 2019`,
    text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
      The building is green and from 18th century.`
  },
  {
    offerId: `2`,
    name: `Bob`,
    rate: 4,
    date: `May 2017`,
    text: `Comment #2`
  },
  {
    offerId: `2`,
    name: `Scarlett`,
    rate: 2,
    date: `May 2018`,
    text: `Comment #3`
  },
  {
    offerId: `3`,
    name: `Bruce`,
    rate: 3,
    date: `June 2020`,
    text: `Comment #4`
  },
  {
    offerId: `3`,
    name: `Amanda`,
    rate: 4,
    date: `January 2018`,
    text: `Comment #5`
  },
  {
    offerId: `3`,
    name: `Maria`,
    rate: 5,
    date: `March 2011`,
    text: `Comment #6`
  }
];
