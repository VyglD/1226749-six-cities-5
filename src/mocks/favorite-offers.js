import offers, {getRandomSubArray} from "./offers";

const MAX_FAVORITE_OFFERS = 3;

export default getRandomSubArray(offers)
  .slice(0, MAX_FAVORITE_OFFERS)
  .map((offer) => offer.id);
