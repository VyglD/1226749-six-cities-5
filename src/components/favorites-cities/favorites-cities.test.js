import React from "react";
import {Provider} from 'react-redux';
import {MemoryRouter} from "react-router-dom";
import renderer from "react-test-renderer";
import {FavoritesCities} from "./favorites-cities";
import {mockFunction, mockOffers, mockOffersFavoriteIds} from "../../mocks/mock-data";
import {mockStore} from "../../mocks/mock-store";

const getFavoriteOffersByCities = (offers, offersIds) => {
  const favoriteOffersByCities = new Map();

  offers.filter((offer) => offersIds.includes(offer.id))
  .forEach((offer) => {
    if (favoriteOffersByCities.has(offer.city)) {
      favoriteOffersByCities.get(offer.city).push(offer);
    } else {
      favoriteOffersByCities.set(offer.city, [offer]);
    }
  });

  return favoriteOffersByCities;
};

it(`Render correctly component FavoritesCities`, () => {
  const component = renderer.create(
      <Provider store={mockStore}>
        <MemoryRouter>
          <FavoritesCities
            favoriteOffersByCities={
              getFavoriteOffersByCities(mockOffers, mockOffersFavoriteIds)
            }
            showCityOffers={mockFunction}
          />
        </MemoryRouter>
      </Provider>
  ).toJSON();

  expect(component).toMatchSnapshot();
});
