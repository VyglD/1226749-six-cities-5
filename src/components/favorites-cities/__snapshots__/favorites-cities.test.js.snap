// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render correctly component FavoritesCities 1`] = `
<ul
  className="favorites__list"
>
  <li
    className="favorites__locations-items"
  >
    <div
      className="favorites__locations locations locations--current"
    >
      <div
        className="locations__item"
      >
        <a
          className="locations__item-link"
          href="#"
        >
          <span>
            Paris
          </span>
        </a>
      </div>
    </div>
    <div
      className="favorites__places"
    >
      <article
        className="place-card favorites__card"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="place-card__mark"
        >
          <span>
            Premium
          </span>
        </div>
        <div
          className="place-card__image-wrapper favorites__image-wrapper"
        >
          <a
            href="/offer/1"
            onClick={[Function]}
          >
            <img
              alt="Place image"
              className="place-card__image"
              height={110}
              src="img/photo1"
              width={150}
            />
          </a>
        </div>
        <div
          className="place-card__info favorites__card-info"
        >
          <div
            className="place-card__price-wrapper"
          >
            <div
              className="place-card__price"
            >
              <b
                className="place-card__price-value"
              >
                €
                111
              </b>
              <span
                className="place-card__price-text"
              >
                / night
              </span>
            </div>
            <button
              className="button place-card__bookmark-button place-card__bookmark-button--active"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="place-card__bookmark-icon"
                height={19}
                width={18}
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="place-card__rating rating"
          >
            <div
              className="place-card__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
          </div>
          <h2
            className="place-card__name"
          >
            <a
              href="/offer/1"
              onClick={[Function]}
            >
              Title of first server offer
            </a>
          </h2>
          <p
            className="place-card__type"
          >
            Apartment
          </p>
        </div>
      </article>
      <article
        className="place-card favorites__card"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
        <div
          className="place-card__image-wrapper favorites__image-wrapper"
        >
          <a
            href="/offer/2"
            onClick={[Function]}
          >
            <img
              alt="Place image"
              className="place-card__image"
              height={110}
              width={150}
            />
          </a>
        </div>
        <div
          className="place-card__info favorites__card-info"
        >
          <div
            className="place-card__price-wrapper"
          >
            <div
              className="place-card__price"
            >
              <b
                className="place-card__price-value"
              >
                €
                1
              </b>
              <span
                className="place-card__price-text"
              >
                / night
              </span>
            </div>
            <button
              className="button place-card__bookmark-button place-card__bookmark-button--active"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="place-card__bookmark-icon"
                height={19}
                width={18}
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="place-card__rating rating"
          >
            <div
              className="place-card__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "60%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
          </div>
          <h2
            className="place-card__name"
          >
            <a
              href="/offer/2"
              onClick={[Function]}
            >
              Title of cheapest offer
            </a>
          </h2>
          <p
            className="place-card__type"
          >
            Hotel
          </p>
        </div>
      </article>
    </div>
  </li>
</ul>
`;
